core/src/main/java/com/linecorp/armeria/client/Endpoint.java

```diff
@@ -418,21 +419,43 @@ public String authority() {
 
         if (isGroup()) {
             authority = "group:" + groupName;
-        } else if (port != 0) {
-            if (hostType == HostType.IPv6_ONLY) {
-                authority = '[' + host() + "]:" + port;
-            } else {
-                authority = host() + ':' + port;
-            }
-        } else if (hostType == HostType.IPv6_ONLY) {
-            authority = '[' + host() + ']';
         } else {
-            authority = host();
+            authority = buildAuthority(host());
         }
 
         return this.authority = authority;
     }
 
+    /**
+     * Converts this endpoint into a URI.
+     *
+     * @return the URI
+     */
+    public URI toURI(String scheme) {
```

Comment:
- `toUri()`
- Could you also add the variants similar to those mentioned in `ClientBuilder`?
  - One thing I'm not sure is whether `Endpoint.toUri(SessionProtocol)` should return a URI that starts with `none+`. Perhaps we should not? :thinking: 
- When a scheme is a `String`, we need to URI-encode it or make sure it does not contain anything weird, e.g. what would happen if `scheme` is `http://www.badguys.com/`?

Reply:
About `http://www.badguys.com`, I find that current URI constructor does not validate `scheme` part correctly, so that `http://www.badguys.com` as scheme is not filtered well.
I can not help but adding `Matcher` for validating `scheme` as described in https://tools.ietf.org/html/rfc3986#section-3.1.